### ISSUE TITLE
Firefox autoraising (all other windows do not)

### ISSUE THREAD

**Issue by pixmin** [koekeishiya/yabai#397]
I have just installed yabai with a minimal setup (to use "focus follow mouse" feature, without auto raise):

```
yabai -m config focus_follows_mouse autofocus
```

It works great, except for Firefox (at least that's the only application I've noticed so far). In addition to getting focused when being mouse over, Firefox also get raised.

Any idea why/who to fix this?

Thanks,

**Comment by koekeishiya** [comment]
The autofocus functionality was implemented by reverse engineering the event-flow of the WindowServer, and so I don't actually know why it will not work for some applications.

Ref. https://github.com/koekeishiya/yabai/issues/333#issuecomment-565710628

**Comment by pixmin** [comment]
If that can help, I just noticed that **PHPStorm** also has the same issue.

**Comment by pixmin** [comment]
And **Compass** (MongoDB) https://www.mongodb.com/products/compass

**Comment by svintit** [comment]
To add here, this seems to be happening to me with any text input windows, they will auto-raise instead of auto-focus. Really believed this was the solution I was looking for but typing in a window that is partially covered by another is the main reason for this.

This happens in both Intellij IDEA and Iterm.

**Comment by cormacrelf** [comment]
Another way the Firefox window difference manifests itself is in `yabai -m window --display X && yabai -m display --focus X`. I'm not sure how related it is, but it's definitely Firefox specific so might help? Anyway:

If I use Safari/most standard windows, I can bounce a window between two displays (next; prev; next; prev) as many times as I like. With Firefox, windows get 'stuck' on a display after a few bounces, i.e. the window moving commands seem to fail. Here's an example I can reproduce on macOS 10.14.6, two displays, yabai is homebrew/2.4.1 but also happens on 2.3.0.

1. `next` ✅ `focus` ✅ (now on display 2)
2. `prev` ✅ `focus` ✅ (now on display 1)
3. `next` ✅ `focus` ✅ (now on display 2)
4. `prev` ❌ (still on display 2)
5. `prev` ❌ (still on display 2)
6. ...

To get it unstuck, you can click on the window, and normal functionality resumes. The window is never unfocused, and the click doesn't refocus it, it just seems to get yabai back in sync. I can get the "n = number of times it works = 1" if the window in question is the only window open on the machine. I.E.

1. `next` ✅ `focus` ✅ (now on display 2)
2. `prev` ❌ (still on display 2)
3. `prev` ❌ (still on display 2)
4. ...

Click on the window, and it will work again.

**Comment by koekeishiya** [comment]
From #333

> there is one application that is refusing to focus but not raise, it's called yEd:

**Comment by guy-dude** [comment]
Add **Simulator** to the list.

**Comment by lucian303** [comment]
Maybe it's related to some cross-platform app tech stack? The ones I see that this happens with are Intellij IDEA, PHPStorm, PyCharm (all Jetbrains IDEs which are java Swing apps and essentially, the same app) as well as Firefox and Blender, both of which use custom GUI toolkits. It could be Swing / other custom GUI toolkits not playing nice. On the other hand other cross platform apps seem fine like Audacity, Bitwig, Visual Studio Code, Slack, etc. Compass, mentioned above, seems to be an electron app that has issues though other electron apps seem ok (Slack, VSC) white yEd is also a cross-platform app. Not sure what the common thing they have is that's preventing autofocus from working though.

**Comment by caduellery** [comment]
Thunderbird (might use the same GUI toolkit as Firefox) is another app that always raise automatically :(

**Comment by lesgrieve** [comment]
Today's Visual Studio Code 1.57.0 update introduced this problem; focussing VSCode causes it to be raised to the front of the desktop. The workaround for that is to downgrade to 1.56.2. I thought this may be of interest since the version introducing the problem could be compared with the previous working version. I opened issue https://github.com/microsoft/vscode/issues/126044 and the response is that the developers plan to fix it in their next update.

**Comment by berenddeboer** [comment]
Yeah, the FireFox issue is really annoying. A big reason I'm only using it in a minor way on a Mac. On Linux it's my main development browser.

**Comment by lesgrieve** [comment]
I just discovered this Bugzilla
https://bugzilla.mozilla.org/show_bug.cgi?id=1707379
Firefox raises window when it gains focus when using yabai with focus_follows_mouse with autofocus, but not autoraise

**Comment by lesgrieve** [comment]
In https://bugzilla.mozilla.org/show_bug.cgi?id=1707379, a patch has been provided for Firefox. Follow https://firefox-source-docs.mozilla.org/setup/macos_build.html to acquire source, apply the patch before `.mach build` step.

**Comment by lesgrieve** [comment]
Today's Firefox Nightly build 97.0a1 (2022-01-07) available on https://www.mozilla.org/en-CA/firefox/channel/desktop/ has this bug fixed. I plan to use this build until the fix is released in the stable version.

**Comment by rickyelopez** [comment]
I have this issue with Microsoft Teams (but only the main chat window, not the call in-progress ' window) and Mattermost. Maybe something to do with electron apps? Someone above mentioned vscode which I think is also an electron app?

**Comment by caduellery** [comment]
Having this issue with Microsoft Teams too, which is annoying, since I have to work with it all day.

**Comment by mcaay** [comment]
I have a workaround that works for Thunderbird so maybe for other apps as well. I noticed that stopping yabai and starting it again when Thunderbird is open fixes it somehow, that yabai can again control the thunderbird window.

My workaround is an alias in my .zshrc file:
```
alias ry='yabai --stop-service && yabai --start-service'
```

So basically open terminal and type `ry`. Not the best solution, but since I rarely shut down my mac or thunderbird it's possible to live with it. If it was needed more often I would put it as a keyboard shortcut in `skhd`.
